apps/store/src/services/Tracking/experimentMiddleware.ts

```diff
@@ -0,0 +1,82 @@
+import { NextRequest, NextResponse } from 'next/server'
+import { EXPERIMENT_COOKIE_NAME, type ExperimentVariant } from './experiment.constants'
+import { getExperimentVariant, getCurrentExperiment } from './experiment.helpers'
+
+const ONE_WEEK = 7 * 24 * 3600
+
+type AssignedVariant = {
+  variant: ExperimentVariant
+  cookieValue: string
+}
+
+export const assignVariant = (): AssignedVariant | undefined => {
+  const currentExperiment = getCurrentExperiment()
+  if (!currentExperiment) return
+
+  let number = Math.random() * 100
+
+  const variant = currentExperiment.variants.find((variant) => {
+    if (variant.weight >= number) return true
+    number -= variant.weight
+  })
+
+  if (!variant) {
+    console.warn(`No variant found for experiment: ${currentExperiment.name}`)
+    return
+  }
+
+  return {
+    variant,
+    cookieValue: `${currentExperiment.id}.${variant.id}`,
+  }
+}
+
+export const getAssignedVariant = (req: NextRequest): AssignedVariant | undefined => {
+  const cookie = req.cookies.get(EXPERIMENT_COOKIE_NAME)
+  if (!cookie) return
+
+  const experimentVariant = getExperimentVariant(cookie.value)
+  if (!experimentVariant) return
+
+  return {
+    variant: experimentVariant,
+    cookieValue: cookie.value,
+  }
+}
+
+export const experimentMiddleware = (req: NextRequest) => {
+  const currentExperiment = getCurrentExperiment()
+  if (!currentExperiment) return
+
+  const slug = currentExperiment.slug.replace(req.nextUrl.locale, '')
+  if (req.nextUrl.pathname !== slug) return
+
+  let assignedVariant = getAssignedVariant(req)
+
+  if (!assignedVariant) {
+    assignedVariant = assignVariant()
```

Comment:
Can this logic of creating and assign a variant in case we don't have one already be placed inside `getAssignedVariant`? Then we could simplify here with:

```suggestion
  if (!assignedVariant) return
```